src/Juvix/Core/IR/Types.hs

```diff
@@ -1,58 +1,140 @@
+{-# LANGUAGE ConstraintKinds, EmptyCase, TypeFamilies, UndecidableInstances #-}
```

Comment:
Can we avoid `UndecidableInstances` or is it required for this pattern?

Reply:
The error message specifically cites nested constraints—e.g. the context for `Eq Term' _ _ _` expands out to:

```haskell
  ( Eq primTy, Eq primVal,
    ( Eq (XStar ext primTy primVal),
      Eq (XPrimTy ext primTy primVal),
      -- etc
    ),
    ( Eq (XBound ext primTy primVal), {- etc -} )
  )
```

...[which is not allowed in instance contexts (or superclasses)](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-ConstraintKinds). It'd probably be possible not to use `UndecidableInstances`, at the cost of more duplication, but I'd like to avoid that unless you feel particularly strongly.

Since this particular restriction only applies to classes & instances, I think this is the only module that will need `UndecidableInstances` enabled.

Reply:
I see - it's not the end of the world - agree that avoiding duplication is also important..

Reply:
btw, since "extension-polymorphic" functions also involve classes, IR.Evaluator needs UIs as well now.